### PR TITLE
ROC-5705 Print legal order and cover sheet on legal order drawn

### PR DESCRIPTION
**Pull request by simad**
**Before creating a pull request make sure that:**

- [x] commit messages are meaningful and follow good commit message guidelines
- [ ] README and other documentation has been updated / added (if needed)
- [x] tests have been updated / new tests has been added (if needed)

Please remove this line and everything above and fill the following sections:

### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/ROC-5705 

### Change description ###

Sending out a letter (via bulk print) with the legal order and the cover sheet to the claimant and to the defendant when a legal order is drawn.

Changed the current bulk print service to accept a list of documents. Considering one of the document we wanted to print is already a PDF - as opposed to a template we create off of html, I added a decorator to print documents. This will allow us to add new document types as we go on.

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/cmc/claimstore/documents/LegalOrderBulkPrintService.java [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    }
+
+    @Recover
+    public void showErrorForBulkPrintFailure(
+        RuntimeException exception,
+        Claim claim,
+        Map<ClaimDocumentType, Document> documents
+    ) {
+        appInsights.trackEvent(BULK_PRINT_FAILED, REFERENCE_NUMBER, claim.getReferenceNumber());
+        throw new RuntimeException("nope");
```
[the remark this answers is not in the document]
Comment (simad) [reply src/main/java/uk/gov/hmcts/cmc/claimstore/documents/LegalOrderBulkPrintService.java]: Ha, I forgot to remove this all together :D

[the remark this answers is not in the document]
Comment (simad) [reply src/main/java/uk/gov/hmcts/cmc/claimstore/documents/LegalOrderBulkPrintService.java]: 😳

File: src/main/java/uk/gov/hmcts/cmc/claimstore/documents/LegalOrderBulkPrintService.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    /* This is configured on Xerox end so they know its us printing and controls things
+     like paper quality and resolution */
+    private static final String XEROX_TYPE_PARAMETER = "CMC001";
+
+    private final SendLetterApi sendLetterApi;
+    private final AuthTokenGenerator authTokenGenerator;
+    private final AppInsights appInsights;
+
+    @Autowired
+    public LegalOrderBulkPrintService(
```
[the remark this answers is not in the document]
Comment (simad) [reply src/main/java/uk/gov/hmcts/cmc/claimstore/documents/LegalOrderBulkPrintService.java]: they differ in how we recover. I'll probably create a common service and override the recover in two sub services

File: src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/DrawOrderCallbackHandler.java [hunk 3]
```diff
@@ -48,6 +58,29 @@ public DrawOrderCallbackHandler(
...
+        SubmittedCallbackResponse.SubmittedCallbackResponseBuilder builder =
+            SubmittedCallbackResponse.builder();
+
+        try {
+            legalOrderService.print(
+                authorisation,
+                caseDetails,
+                draftOrderDoc);
+        } catch (Exception e) {
+            builder.confirmationHeader("Bulk Print Failed");
```
[the remark this answers is not in the document]
Comment (simad) [reply src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/DrawOrderCallbackHandler.java]: afaik those errors will be showed as a banner on the ccd ui so they will have the case open right in front of their eyes :D

File: src/main/java/uk/gov/hmcts/cmc/claimstore/services/staff/content/legaladvisor/LegalOrderCoverSheetContentProvider.java [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+            claim.getClaimData().getDefendant().getName(),
+            claim.getClaimData().getDefendant().getAddress()
+        );
+    }
+
+    private Map<String, Object> createContent(Claim claim, String partyFullName, Address partyAddress) {
+        requireNonNull(claim);
+
+        Map<String, Object> content = new HashMap<>();
+        content.put("partyFullName", partyFullName);
```
[the remark this answers is not in the document]
Comment (simad) [reply src/main/java/uk/gov/hmcts/cmc/claimstore/services/staff/content/legaladvisor/LegalOrderCoverSheetContentProvider.java]: 👍

File: src/main/java/uk/gov/hmcts/cmc/claimstore/services/staff/content/legaladvisor/LegalOrderCoverSheetContentProvider.java [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+    public Map<String, Object> createContentForDefendant(Claim claim) {
+        return createContent(
+            claim,
+            claim.getClaimData().getDefendant().getName(),
+            claim.getClaimData().getDefendant().getAddress()
+        );
+    }
+
+    private Map<String, Object> createContent(Claim claim, String partyFullName, Address partyAddress) {
```
[the remark this answers is not in the document]
Comment (simad) [reply src/main/java/uk/gov/hmcts/cmc/claimstore/services/staff/content/legaladvisor/LegalOrderCoverSheetContentProvider.java]: yep. originally I thought we needed more information from claim I'll change that 👍

File: src/main/java/uk/gov/hmcts/cmc/claimstore/events/claim/PinOrchestrationService.java [hunk 6]
```diff
@@ -46,8 +48,12 @@ public Claim process(Claim claim, String authorisation, String submitterName) {
             ClaimSubmissionOperationIndicators.builder();
 
         try {
-            bulkPrintService.print(updatedClaim, documents.getDefendantPinLetterDoc(),
-                documents.getSealedClaimDoc());
+            bulkPrintService.print(
+                updatedClaim,
+                ImmutableMap.of(
+                    ClaimDocumentType.DEFENDANT_PIN_LETTER, documents.getDefendantPinLetterDoc(),
```
[the remark this answers is not in the document]
Comment (simad) [reply src/main/java/uk/gov/hmcts/cmc/claimstore/events/claim/PinOrchestrationService.java]: I'll change that 👍

File: src/main/java/uk/gov/hmcts/cmc/claimstore/services/staff/content/legaladvisor/LegalOrderService.java [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+        DocumentManagementService documentManagementService,
+        CCDCaseDataToClaim ccdCaseDataToClaim) {
+        this.documentTemplates = documentTemplates;
+        this.legalOrderBulkPrintService = legalOrderBulkPrintService;
+        this.legalOrderCoverSheetContentProvider = legalOrderCoverSheetContentProvider;
+        this.documentManagementService = documentManagementService;
+        this.ccdCaseDataToClaim = ccdCaseDataToClaim;
+    }
+
+    private Document downloadLegalOrder(String authorisation, CCDDocument ccdLegalOrder) throws URISyntaxException {
```
Comment (dharmendrak) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/services/staff/content/legaladvisor/LegalOrderService.java]: We should have first public method followed by private's?

File: src/main/java/uk/gov/hmcts/cmc/claimstore/utils/PartyNameUtils.java [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+
+    private PartyNameUtils() {
+        // Utility class, no instances
+    }
+
+    public static String getPartyNameFor(NamedParty party) {
+        StringBuilder name = new StringBuilder(party.getName());
+
+        if (party instanceof SoleTrader) {
+            (((SoleTrader) party).getBusinessName()).ifPresent(t -> name.append(" T/A ").append(t));
```
Comment (dharmendrak) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/utils/PartyNameUtils.java]: is this business requirement?

File: src/main/java/uk/gov/hmcts/cmc/claimstore/services/staff/content/legaladvisor/LegalOrderService.java (hunk 7 above)
Comment (simad) [reply src/main/java/uk/gov/hmcts/cmc/claimstore/services/staff/content/legaladvisor/LegalOrderService.java]: sure 👍

File: src/main/java/uk/gov/hmcts/cmc/claimstore/utils/PartyNameUtils.java (hunk 8 above)
Comment (simad) [reply src/main/java/uk/gov/hmcts/cmc/claimstore/utils/PartyNameUtils.java]: not in this story, but that's what we already had for other letters addressed to the claimant. Will check with Sanchita

**Pull request merged**